### PR TITLE
Pass principal type during role assignment put operations

### PR DESCRIPTION
**Pull request by dagoroz**
## Description

Added checks to ObjectPrincipal type to see if it's included in the PUT request, if not it adds it automatically

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.cs [hunk 1]
```diff
@@ -174,7 +174,7 @@ public void RaGetByUPNWithExpandPrincipalGroups()
             TestRunner.RunTestScript("Test-RaGetByUPNWithExpandPrincipalGroups");
         }
 
-        [Fact]
+        [Fact(Skip = "Breaking PR in 2 parts needs this test to be disabled temporarily")]
```
Comment (dagoroz) [inline src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.cs:177]: The change I'm making changes the codeflow intended for when creating role assignments through an SP (e.g. Azure app)
The flow is not broken but it does need to be updated. That will be done on a separate PR as it might be an extensive change by itself.

File: src/Resources/Resources/Models.Authorization/AuthorizationClient.cs [hunk 2]
```diff
@@ -128,7 +128,15 @@ public IEnumerable<PSRoleDefinition> FilterRoleDefinitionsByCustom(string scope,
         /// <returns>The created role assignment object</returns>
         public PSRoleAssignment CreateRoleAssignment(FilterRoleAssignmentsOptions parameters, Guid roleAssignmentId = default(Guid))
         {
-            string principalId = ActiveDirectoryClient.GetObjectId(parameters.ADObjectFilter);
+            PSADObject asignee = ActiveDirectoryClient.GetADObject(parameters.ADObjectFilter);
+
+            if (asignee == null)
+            {
+                throw new ArgumentException("No AD object could be found with current parameters, please confirm the information provided is correct and try again");
```
Comment (isra-fel) [inline src/Resources/Resources/Models.Authorization/AuthorizationClient.cs:135]: Please use a resource string instead

File: src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.ps1 [hunk 3]
```diff
@@ -707,17 +707,19 @@ function Test-RaCreatedBySP
     #Setup
     # Conect to azure with SP
     # If you need to re-record replace this setup
-    $passwd = ConvertTo-SecureString 'MCg1SVb3MX' -AsPlainText -Force
-    $pscredential = New-Object System.Management.Automation.PSCredential('c82180ed-2f4b-4cb2-965f-ec5088751710' , $passwd)
-    $tenantId = '1462fd46-afe5-491b-a340-31ebae81d1ce'
-    Connect-AzAccount -ServicePrincipal -Credential $pscredential -Tenant $tenantId
+    $passwd = ConvertTo-SecureString 'm7XHgb9cDVP60L4.3bT~VN1Rhjpo-.ZTU-' -AsPlainText -Force
```
Comment (isra-fel) [inline src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.ps1:710]: Just want to double check if the credential/SP has been removed

File: src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.cs (hunk 1 above)
Comment (isra-fel) [reply src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.cs:177]: Sounds OK, but what about the above cases with `Skip = "Test is not deterministic, only works on record mode"`?

**Review by isra-fel — CHANGES_REQUESTED** [review]
Generally LGTM. Added a few comments

File: src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.cs (hunk 1 above)
Comment (dagoroz) [reply src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.cs:177]: the tests themselves pass, in record mode they use my credentials and my tenant and everything goes well ✅
the problem is that the tests generate guids on the fly which break the tests on playback mode. When trying to read the guid from the session records, since it always change it's impossible to request the same guid twice.

File: src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.ps1 (hunk 3 above)
Comment (dagoroz) [reply src/Resources/Resources.Test/ScenarioTests/RoleAssignmentTests.ps1:710]: I removed the credentials and the SP from subscription

File: src/Resources/Resources/Models.Authorization/AuthorizationClient.cs (hunk 2 above)
Comment (dagoroz) [reply src/Resources/Resources/Models.Authorization/AuthorizationClient.cs]: changed

**Review by isra-fel — APPROVED** [review]
